### PR TITLE
Support HEIF image file

### LINKED ISSUE #1021: heif format / .heic image file support
Are there any plans to support .heic image files? For pillow there is https://pypi.org/project/heif-image-plugin/ based on https://github.com/carsales/pyheif.

### PR DESCRIPTION
**Pull request by Pandede**
Resolves #1021.
The previous pull request #1040 is closed as the redundant commits.
And also, move `pillow-heif` to be a plugin rather than extra requirement of `Pillow` (thanks for the advice of @bigcat88).

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -99,6 +99,7 @@
     "swf": [],
     "tifffile": ["tifffile"],
     "pyav": ["av"],
+    "heif": ["pillow-heif"],
```
Comment (bigcat88) [inline setup.py:102]: should be:
```
"pillow-heif": [],
```

**Comment by bigcat88** [comment]
never mind about my previous deleted comment.

as I see, CI needs wheels for PyPy3.8 - if I will know that this will have high chances to be merged, I can restore the publishing of PyPy3.8 wheels for the next release, if it is needed.

**Review by FirefoxMetzger — CHANGES_REQUESTED** [review]
Awesome initiative, thanks @Pandede 🚀 This looks pretty good from the plugin implementation side. I think we'll only need one round of reviews before we can merge.

I think we should also properly register the format so that pillow is automatically tried first for `.heic` files. To do so we need to add a new [`FileExtension`](https://github.com/imageio/imageio/blob/0bbf1962c7c36ebee28639aee8441e01d4168215/imageio/config/extensions.py#L9) in the relevant [config file](https://github.com/imageio/imageio/blob/0bbf1962c7c36ebee28639aee8441e01d4168215/imageio/config/extensions.py#L1854-L1857). This will also add the necessary documentation so that HEIC shows up in the list of supported formats :)

Beyond that, there is one change (making `pillow_heif` option) that I would like to see before we merge.

**Comment by bigcat88** [comment]
> as I see, CI needs wheels for PyPy3.8 - if I will know that this will have high chances to be merged, I can restore the publishing of PyPy3.8 wheels for the next release, if it is needed.

done from my side, new release was published :)

File: tests/test_pillow.py [hunk 2]
```diff
@@ -705,3 +705,10 @@ def test_writable_output():
 
     frame = iio.imread(buffer, writeable_output=False, plugin="pillow")
     assert not frame.flags["WRITEABLE"]
+
+
+@pytest.mark.needs_internet
+def test_heif_remote():
+    url = "https://nokiatech.github.io/heif/content/images/surfer_1440x960.heic"
```
[the remark this answers is not in the document]
Comment (Pandede) [reply tests/test_pillow.py:712]: I may use the HEIC images in this repository instead: https://github.com/tigranbs/test-heic-images.

**Comment by Pandede** [comment]
The tests are failed as the latest upgrade of `Pillow`. All tests passed if `Pillow==10.0.1` but failed if `Pillow==10.1.0`.

**Comment by FirefoxMetzger** [comment]
Yes. There is another issue related to pillow's recent upgrade. I'll pin pillow for now and then see how we can best resolve this.

**Review by FirefoxMetzger — APPROVED** [review]
Looks good from my side.

I will check why why the pypi CI no longer likes the av package and then we should be able to get a green CI :)

File: imageio/config/extensions.py [hunk 3]
```diff
@@ -1855,6 +1855,9 @@ def reset(self):
         extension=".btf",
         priority=["tifffile"],
     ),
+    FileExtension(
+        name="High Efficiency Image File Format", extension=".heic", priority=["pillow"]
```
[the remark this answers is not in the document]
Comment (FirefoxMetzger) [reply imageio/config/extensions.py:1859]: @kmilos Looks like @Pandede is one step ahead: https://github.com/imageio/imageio/pull/1048

AVIF support is already merged :) as for HIF I'm not sure. Is it just a different suffix for HEIF but actually the same format?

[the remark this answers is not in the document]
Comment (kmilos) [reply imageio/config/extensions.py:1859]: Yes. Some cameras now output .HIF directly (I guess limitations of sdcard  filesystem like old 8.3 filename format from DOS times!)

[the remark this answers is not in the document]
Comment (kmilos) [reply imageio/config/extensions.py:1859]: 2.33.0 is still missing ".heif" and ".hif".

Also, libheif from 1.17 supports the JPEG 2000 codec inside the HEIF container w/ a ".hej2" extension, see https://www.itu.int/rec/T-REC-T.815/en

[the remark this answers is not in the document]
Comment (Pandede) [reply imageio/config/extensions.py:1859]: Maybe create a new issue about the multiple extensions of HEIF?

**Pull request merged**